### ISSUE TITLE
0EB$ doesn't show buffer ring

### ISSUE THREAD

**Issue by blakemcbride** [rhaberkorn/sciteco#15]
0EB$ doesn't show buffer ring
Instead, it is creating an unnamed buffer.

**Comment by rhaberkorn** [comment]
The buffer ring is shown when you type `0EB` **until** you hit Escape. So to just query the buffer ring, you type `0EB` and rub it out afterwards. The 0 *additionally* shows the buffer ring but otherwise does not alter the behaviour of `EB`, so you can show the buffer ring and immediately go on opening some other file.

All in all, I am not that happy with the current behaviour - inherited from Video TECO - but I am not sure how to do it better. Perhaps, there should simply be a special immediate editing command to autocomplete from the list of files in the buffer ring (if the string argument is still empty) instead of from the files in the current directory.

**Comment by blakemcbride** [comment]
0EB lists as:

<absolut-path-1>  <absoulte-path-2>  <absolute-path-3>

I'd rather see:

1 <absolute-path-1>
2 <absolute-path-2>
3 <absolute-path-3>

Also, in teh case of 0EB, I would make 0EB$ do the same thing.  In other words, ignore that trailing $.

The problem is that the EB command normally taks an argument (the file name).  It doesn't normally react until the trailing $.  It wouldn't be possible.  For consistency sake, I think it would be better if 0EB requires $.

Thanks!

**Comment by rhaberkorn** [comment]
> I'd rather see:
> 
> 1 <absolute-path-1> 2 <absolute-path-2> 3 <absolute-path-3>

That's a matter of taste. But it would definitely make sense to see the buffer id prepended to the path. I will add that to my TODO. Of course you can also select the absolute path by mouse and insert it back from the clipboard to open a particular file from the ring. So for instance, you type `0EB`, see what file you want, select it, insert it (middle mouse button in Linux) and hit Escape.

> For consistency sake, I think it would be better if 0EB required $.

That's definitely not going to happen since `EB$` is already taken to open the unnamed buffer. It would be inconsistent to not do that when prepended with a 0 as long as it will open any file even in the presence of 0. But see above, it would probably be best to abolish `0EB` altogether and allow to autocomplete from the buffer ring. Unfortunately, we cannot use Shift-Tab in Curses, so you would have to use something like Ctrl+G (enable immediate editing modifier), followed by Tab. Or some unrelated function key. Or some Ctrl+Letter combination would also be possible. This autocompletion would only make sense in filename parameters, so you don't loose much if you cannot easily insert some ASCII character smaller than 32 (and they can always be inserted using Caret+Letter combinations).

**Comment by blakemcbride** [comment]
The reason I like them vertical rather than horizontal is becase they're easy to compare.  If they're in different directories or if their spelling is slightly off, it's easty to see if they are displayed vertically.  If they're displayed horizontally like they are now, it is very difficult to see their differences.

**Comment by rhaberkorn** [comment]
But ordinary file completion popups you would like to keep as they are?

**Comment by blakemcbride** [comment]
I hadn't seen those yet.

**Comment by blakemcbride** [comment]
Having file completion popups displayed horizontally is much less of an issue I would think becasue they're all in the same directory and (I presume) alphbetical.

**Comment by blakemcbride** [comment]
IMO, having <code>0EB$</code> open an unnamed buffer makes sense and is consistent.  However, I don't think having <code>0EB</code> display the buffer list is good.  <code>EB</code> is being inconsistently overloaded.  I'd suggest using a differend command sequence to display the buffer list.

**Comment by rhaberkorn** [comment]
> However, I don't think having `0EB` display the buffer list is good.

Agreed, especially since language-wise 0EB does nothing. A new command would also be a possibility. But often when you look at the buffer list, you actually want to open a file afterwards, so having it somehow integrated into EB does make some sense. Considering how SciTECO currently works, a special completion command like Ctrl+T or Ctrl+B or whatever would make most sense IMHO.

**Comment by blakemcbride** [comment]
Agreed.  How about this -

<code>XX</code> (XX=whatever) displays the buffer list.  <code>XXn$</code> switched to buffer number <code>n</code>?

This way you can display the buffer list and simple put in the buffer number and <code>$</code> and you've got it!  No need to back up!

**Comment by rhaberkorn** [comment]
Note that with my most recent changes, you can also click on popup entries to complete them. So you can type `0EB` and just click on any entry to switch to that file. On ncurses, this requires enabling mouse support with `0,64ED` - but that's the new default in sample.teco_ini.

Still, a special control-key to show that same popup would be more consistent. Considering this feature, would it still be useful to show the buffer ids in the popups? Probably not.

**Comment by blakemcbride** [comment]
With all of these hits (e.g.  "that's not going to happen because XYX is already used for ABC), I propose the following:

Use standard TECO commands 100%.  Duplicate them as closely as possible.  Add your features on top.  This way your editor ends up being a significant addition to the TECO line rather than a thing unto itself.  People who know TECO can use it right off the bat and learn your extras over time similar to VI -> VIM.

It is my opinion that this approach would lead to the greatest acceptance and use of SciTeco.

**Comment by rhaberkorn** [comment]
> Use standard TECO commands 100%. Duplicate them as closely as possible. Add your features on top. This way, your editor ends up being a significant addition to the TECO line rather than a thing unto itself. People who know TECO can use it right off the bat and learn your extras over time, similar to VI -> VIM.

Frankly speaking, I don't think that's a good idea.

- Many features of "Standard" TECO (TECO-11) do not make sense on modern machines and workflows and are just wasted keyboard real estate. Stuff that's useful and in demand on modern screen editors however, would have to be added with longer and more cumbersome commands.
- Supporting all of "standard" TECO adds unnecessary code cruft as I would have to implement many features, most of them I would myself never ever use. This would go as far as having to add classic REPL modes without Undo, classic screen modes etc.
- The "Standard" TECO parser has serious shortcomings. I do not want to keep strict compatibility to it.
- Almost no one uses and knows DEC TECO anymore. It's not that we had a large user base of people fluent in TECO-11, that would be worth to pick up.
- While DEC TECO is now the most well known dialect of TECO and the one cloned in most other "modern" implementations, it's by far the only one. Historically, at the heyday of TECO, there have been dozen of dialects. Practically everybody forked his own version and changed the language often in much more radical ways than I did. I recommend you to read old editions of "The Moby Munger". I think I am moving well within that realm.
- I am not a Vi expert, but as far as I see, Vi implementations also differ a lot. They are not strict supersets of historic Vi. They all just share the basic approach and most of the basic keyboard shortcuts, so they are recognizable Vis.

So what I decided to do, is to strive for TECO-11 compatibility, where it does not significantly hurt the usability of the editor, but not to slavishly restrict myself to a superset of it. Also to be open for new ideas and concepts, that just wouldn't be possible in classic TECOs. At the same time, I'd like that the editor and language stays recognizably a TECO. With that approach, I am already going beyond what would make sense if I were to design a language from scratch, e.g. in the way that Q-Register cells are managed. The language is close enough to TECO 11 so that scripts like [PI.TEC](https://github.com/rhaberkorn/sciteco/wiki/Calculating-Pi) work with minimal modifications. TECO 11 compatibility will further improve until the v3.0 release. TECO 11 still is the main influence in SciTECO.

SciTECO is not to be understood like FORTRAN 90, compared to FORTRAN 77 or QBasic compared to ANSI BASIC. It's more like Visual Basic compared to ANSI BASIC. Not a very flattering comparison, I know.
